plugin-server/src/cdp/hog-watcher/hog-watcher.ts

```diff
@@ -0,0 +1,472 @@
+
+                observations.forEach(({ id, observation }) => {
+                    const items = (this.globalState!.observations[id] = this.globalState!.observations[id] ?? [])
+                    items.push({
+                        ...observation,
+                        instanceId: instanceId,
+                    })
+                })
+            },
+
+            'hog-watcher-states': (message) => {
+                // NOTE: This is only emitted by the leader so we can immediately add it to the list of states
+                const { states }: EmittedHogWatcherStates = JSON.parse(message)
+
+                this.states = {
+                    ...this.states,
+                    ...states,
+                }
+            },
+        })
+    }
+
+    async start() {
+        await this.pubSub.start()
+
+        // Get the initial state of the watcher
+        await this.syncStates()
+
+        if (process.env.NODE_ENV === 'test') {
+            // Not setting up loop in test mode
+            return
+        }
+
+        const loop = async () => {
+            try {
+                // Maybe add a slow function warning here
+                await this.sync()
+            } finally {
+                this.interval = setTimeout(loop, OBSERVATION_PERIOD)
+            }
+        }
+
+        await loop()
+    }
+
+    async stop() {
+        await this.pubSub.stop()
+
+        if (this.interval) {
+            clearTimeout(this.interval)
+        }
+        if (!this.isLeader) {
+            return
+        }
+
+        await runRedis(this.hub.redisPool, 'stop', async (client) => {
+            return client.del(`${BASE_REDIS_KEY}/leader`)
+        })
+
+        await this.flushActiveObservations()
+    }
+
+    public getFunctionState(id: HogFunctionType['id']): HogWatcherState {
+        return this.states[id] ?? HogWatcherState.healthy
+    }
+
+    private async checkIsLeader() {
+        const leaderId = await runRedis(this.hub.redisPool, 'getLeader', async (client) => {
+            // Set the leader to this instance if it is not set and add an expiry to it of twice our observation period
+            const pipeline = client.pipeline()
+
+            // TODO: This can definitely be done in a single command - just need to make sure the ttl is always extended if the ID is the same
+
+            // @ts-expect-error - IORedis types don't allow for NX and EX in the same command
+            pipeline.set(`${BASE_REDIS_KEY}/leader`, this.instanceId, 'NX', 'EX', (OBSERVATION_PERIOD * 3) / 1000)
+            pipeline.get(`${BASE_REDIS_KEY}/leader`)
+            const [_, res] = await pipeline.exec()
+
+            // NOTE: IORedis types don't allow for NX and GET in the same command so we have to cast it to any
+            return res[1] as string
+        })
+
+        this.isLeader = leaderId === this.instanceId
+
+        if (this.isLeader) {
+            status.info('👀', '[HogWatcher] I am the leader')
+        }
+    }
```

Comment:
I wonder... if we should not just have an explicitly deployed leader? This would remove all that coordination noise, and give it breathing room from running hog code as well.

This happened before. The plugin server has a [scheduler](https://github.com/PostHog/posthog/blob/30bafdd38278b3fb6d20f53984df3b0154cf4a37/plugin-server/src/main/graphile-worker/schedule.ts), which used something called [redlock](https://github.com/PostHog/posthog/blob/e5d814b4bae6fdb803464fcac60c49732591034a/plugin-server/src/utils/redlock.ts) to make sure there's only one pod in the fleet running the scheduling commands. This made sense when there were 2 scheduled tasks running per hour on a self hosted instance, but the Cloud required a different approach, as these scheduled bursts in ingestion nodes were causing problems. 

Now we have a lot of schedulers:

![image](https://github.com/PostHog/posthog/assets/53387/fb78bc57-44d4-4110-9baf-829230bc527d)

Thus, I think we'd make this system more robust if we'd remove the redis lock and make it just a single node service. We would immediately buy some vertical scaling room as well.

Reply:
I think this is a good idea tbh. I didn't do it as there was pushback on the leader idea in general and I didn't want to go too far down that whole if there might have been an alternative I realised along the way.

Should be easy enough to setup, but again I might do that in follow up as its more of an improvmenet